Fuzzing-sim causing special order to cross

# Problem encountered
Market-sim [fuzzing scenario](https://github.com/vegaprotocol/vega-market-sim/pull/347) runs multiple markets with sensible and degenerate traders and LPs who trigger multiple close-outs. Fuzzing agents submit batches of random order submission, amendments, and cancellations to try and cause panics.

Market-sim was running latest core release ([v0.68.3](https://github.com/vegaprotocol/vega/releases/tag/v0.68.3))

# Observed behaviour
Simulation ran until special orders crossed and panic raised.
```
panic: re submit special order would cross

goroutine 65570 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x0?, 0x2?, {0x2?, 0x0?, 0x0?})
	/Users/caed/go/pkg/mod/go.uber.org/zap@v1.23.0/zapcore/entry.go:198 +0x7c
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0x1400e19ab60, {0x14010a0e500, 0x2, 0x2})
	/Users/caed/go/pkg/mod/go.uber.org/zap@v1.23.0/zapcore/entry.go:264 +0x2ac
go.uber.org/zap.(*Logger).Panic(0x44b47da0bbb14000?, {0x106db9b55?, 0x0?}, {0x14010a0e500, 0x2, 0x2})
	/Users/caed/go/pkg/mod/go.uber.org/zap@v1.23.0/logger.go:251 +0x54
code.vegaprotocol.io/vega/core/matching.(*OrderBook).ReSubmitSpecialOrders(0x1400e572640, 0x1401c2cd380)
	/Users/caed/vega-market-sim/extern/vega/core/matching/orderbook.go:807 +0x51c
code.vegaprotocol.io/vega/core/execution.(*Market).reSubmitPeggedOrders(0x140001cc000, {0x107e4bad8, 0x14010b71740}, {0x14010ab2940, 0x3, 0x0?})
	/Users/caed/vega-market-sim/extern/vega/core/execution/special_orders.go:110 +0xd0
code.vegaprotocol.io/vega/core/execution.(*Market).repriceAllSpecialOrders(0x140001cc000, {0x107e4bad8, 0x14010b71740}, 0xb8?, {0x14010a24340?, 0x107d6a8e0?, 0x14010b71701?}, 0x107e4bad8?, 0x14010a24268?)
	/Users/caed/vega-market-sim/extern/vega/core/execution/special_orders.go:165 +0x2d4
code.vegaprotocol.io/vega/core/execution.(*Market).checkForReferenceMoves(0x140001cc000, {0x107e4bad8, 0x14010b71740}, {0x14010a24340, 0x0, 0x0}, 0x1)
	/Users/caed/vega-market-sim/extern/vega/core/execution/reference_price_moves.go:80 +0x2b4
code.vegaprotocol.io/vega/core/execution.(*Market).leaveAuction(0x140001cc000, {0x107e4bad8, 0x14010b71740}, {0xf710890?, 0xed936c007?, 0x0?})
	/Users/caed/vega-market-sim/extern/vega/core/execution/market.go:1113 +0x890
code.vegaprotocol.io/vega/core/execution.(*Market).checkAuction(0x140001cc000, {0x107e4bad8, 0x14010b71740}, {0xf710890?, 0xed936c006?, 0x0?})
	/Users/caed/vega-market-sim/extern/vega/core/execution/auction.go:112 +0x6e4
code.vegaprotocol.io/vega/core/execution.(*Market).OnTick(0x140001cc000, {0x107e4bad8, 0x14010b71740}, {0xf2e?, 0x14019d08660?, 0x0?})
	/Users/caed/vega-market-sim/extern/vega/core/execution/market.go:730 +0x5e4
code.vegaprotocol.io/vega/core/execution.(*Engine).OnTick(0x140002fa540, {0x107e4bad8, 0x14010b71740}, {0x1?, 0x107e64858?, 0x0?})
	/Users/caed/vega-market-sim/extern/vega/core/execution/engine.go:778 +0x1a0
code.vegaprotocol.io/vega/core/vegatime.(*Svc).notify(...)
	/Users/caed/vega-market-sim/extern/vega/core/vegatime/service.go:104
code.vegaprotocol.io/vega/core/vegatime.(*Svc).SetTimeNow(0x14000290200, {0x107e4bad8, 0x14010b71740}, {0x1098fc240?, 0x1076c1cc0?, 0x0?})
	/Users/caed/vega-market-sim/extern/vega/core/vegatime/service.go:77 +0x278
code.vegaprotocol.io/vega/core/processor.(*App).OnBeginBlock(_, {{0x14019b4f360, 0x20, 0x20}, {{0x0, 0x0}, {0x14002028688, 0x6}, 0x10ba, {0xf710890, ...}, ...}, ...})
	/Users/caed/vega-market-sim/extern/vega/core/processor/abci.go:781 +0x554
code.vegaprotocol.io/vega/core/blockchain/abci.(*App).BeginBlock(_, {{0x14019b4f360, 0x20, 0x20}, {{0x0, 0x0}, {0x14002028688, 0x6}, 0x10ba, {0xf710890, ...}, ...}, ...})
	/Users/caed/vega-market-sim/extern/vega/core/blockchain/abci/abci.go:46 +0x60
code.vegaprotocol.io/vega/cmd/vega/node.(*appW).BeginBlock(_, {{0x14019b4f360, 0x20, 0x20}, {{0x0, 0x0}, {0x14002028688, 0x6}, 0x10ba, {0xf710890, ...}, ...}, ...})
	/Users/caed/vega-market-sim/extern/vega/cmd/vega/node/app_wrapper.go:64 +0x4c
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).BeginBlock(0x14000644360)
	/Users/caed/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:336 +0x2ac
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).processBlock(0x14000644360)
	/Users/caed/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:195 +0x1b0
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).ForwardTime(0x14000644360, 0x3b9aca00)
	/Users/caed/vega-market-sim/extern/vega/core/blockchain/nullchain/nullchain.go:272 +0x198
code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime.func1()
	/Users/caed/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:129 +0x2c
created by code.vegaprotocol.io/vega/core/blockchain/nullchain.(*NullBlockchain).handleForwardTime
	/Users/caed/vega-market-sim/extern/vega/core/blockchain/nullchain/server.go:128 +0x1ec
```

# Evidence

## Logs
[node.txt](https://github.com/vegaprotocol/vega/files/10908579/node_error_logs.txt)
[replay.txt](https://github.com/vegaprotocol/vega/files/10908593/replay.txt)
[genesis.txt](https://github.com/vegaprotocol/vega/files/10913028/genesis.txt)



## Additional context
Market-sim was running latest core release ([v0.68.3](https://github.com/vegaprotocol/vega/releases/tag/v0.68.3))

# Definition of Done
>ℹ️ Not every issue will need every item checked, however, every item on this list should be properly considered and actioned to meet the [DoD](https://github.com/vegaprotocol/vega/blob/develop/DEFINITION_OF_DONE.md).

**Before Merging**
- [ ] Code refactored to meet SOLID and other code design principles
- [ ] Code is compilation error, warning, and hint free
- [ ] Carry out a basic happy path end-to-end check of the new code
- [ ] All APIs are documented so auto-generated documentation is created
- [ ] All bug recreation steps can be followed without presenting the original error/bug
- [ ] All Unit, Integration and BVT tests are passing
- [ ] Implementation is peer reviewed (coding standards, meeting acceptance criteria, code/design quality)
- [ ] Create [front end](https://github.com/vegaprotocol/token-frontend/issues) or [console](https://github.com/vegaprotocol/console/issues) tickets with feature labels (should be done when starting the work if dependencies known i.e. API changes)

**After Merging**
- [ ] Move development ticket to `Done` if there is **NO** requirement for new [system-tests](https://github.com/vegaprotocol/system-tests/issues)
- [ ] Resolve any issues with broken system-tests
- [ ] Create [documentation](https://github.com/vegaprotocol/documentation/issues) tickets with feature labels if functionality has changed, or is a new feature

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -50,6 +50,7 @@
 - [7499](https://github.com/vegaprotocol/vega/issues/7499) - Implement transaction check functionality to wallet
 - [7745](https://github.com/vegaprotocol/vega/issues/7745) - Use margin after the application of a bond penalty to assess LP solvency
 - [7765](https://github.com/vegaprotocol/vega/issues/7765) - Assure pegged order won't get deployed with insufficient margin
+- [7786](https://github.com/vegaprotocol/vega/issues/7786) - Fix validation of order amendments (check for negative pegged offset)
 
 ## 0.68.0
 

--- a/commands/order_amendment.go
+++ b/commands/order_amendment.go
@@ -90,6 +90,11 @@ func checkOrderAmendment(cmd *commandspb.OrderAmendment) Errors {
 
 	if cmd.PeggedOffset != "" {
 		isAmending = true
+		if peggedOffset, ok := big.NewInt(0).SetString(cmd.PeggedOffset, 10); !ok {
+			errs.AddForProperty("order_amendment.pegged_offset", ErrNotAValidInteger)
+		} else if peggedOffset.Cmp(big.NewInt(0)) <= 0 {
+			errs.AddForProperty("order_amendment.pegged_offset", ErrMustBePositive)
+		}
 	}
 
 	if !isAmending {

--- a/commands/order_amendment_test.go
+++ b/commands/order_amendment_test.go
@@ -24,6 +24,7 @@ func TestCheckOrderAmendment(t *testing.T) {
 	t.Run("amend order invalid expiry type - fail", testAmendOrderInvalidExpiryFail)
 	t.Run("amend order tif to GFA - fail", testAmendOrderToGFA)
 	t.Run("amend order tif to GFN - fail", testAmendOrderToGFN)
+	t.Run("amend order pegged_offset", testAmendOrderPeggedOffset)
 }
 
 func testNilOrderAmendmentFails(t *testing.T) {
@@ -123,6 +124,26 @@ func testAmendOrderInvalidExpiryFail(t *testing.T) {
 	arg.TimeInForce = proto.Order_TIME_IN_FORCE_IOC
 	err = checkOrderAmendment(arg)
 	assert.Error(t, err)
+}
+
+func testAmendOrderPeggedOffset(t *testing.T) {
+	arg := &commandspb.OrderAmendment{
+		OrderId:      "08dce6ebf50e34fedee32860b6f459824e4b834762ea66a96504fdc57a9c4741",
+		MarketId:     "08dce6ebf50e34fedee32860b6f459824e4b834762ea66a96504fdc57a9c4741",
+		PeggedOffset: "-1789",
+		TimeInForce:  proto.Order_TIME_IN_FORCE_GTC, // just here to test the case with empty pegged offset
+	}
+
+	err := checkOrderAmendment(arg)
+	assert.Error(t, err.ErrorOrNil())
+
+	arg.PeggedOffset = ""
+	err = checkOrderAmendment(arg)
+	assert.NoError(t, err.ErrorOrNil())
+
+	arg.PeggedOffset = "1000"
+	err = checkOrderAmendment(arg)
+	assert.NoError(t, err.ErrorOrNil())
 }
 
 /*

